homeassistant/components/decora_wifi/__init__.py

```diff
@@ -1 +1,104 @@
+from decora_wifi.models.permission import Permission
+from decora_wifi.models.person import Person
+from decora_wifi.models.residence import Residence
+from decora_wifi.models.residential_account import ResidentialAccount
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+
+from .config_flow import BASE_SCHEMA
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS = [Platform.LIGHT]
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: BASE_SCHEMA},
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    session = DecoraWiFiSession()
+    user = await hass.async_add_executor_job(lambda: session.login(username, password))
+    if not user:
+        raise ConfigEntryError("could not authenticate")
+
+    hass.data[DOMAIN][entry.entry_id] = session
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    def logout(event: Event) -> None:
+        if session is not None:
+            try:
+                Person.logout(session)
+            except ValueError as err:
+                _LOGGER.error("Failed to log out of myLeviton Service: %s", err)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    hass.data.pop(DOMAIN)
+    return unload_ok
+
+
+class DecoraWifiAsyncClient:
+    """A thin wrapper to make async calls more readable."""
+
+    def __init__(self, session: DecoraWiFiSession, hass: HomeAssistant) -> None:
+        """DecoraWifiAsyncClient."""
+        self.session = session
+        self.hass = hass
+
+    async def get_permissions(self):
+        """Get all permissions for the provided session."""
+        assert self.session.user
+        return await self.hass.async_add_executor_job(
+            self.session.user.get_residential_permissions
+        )
+
+    async def get_residences(self, permissions: list[Permission]):
+        """Get all residences for the provided permissions."""
+        residences: list[Residence] = []
+        for perm in permissions:
+            if perm.residentialAccountId is not None:
+                account = ResidentialAccount(self.session, perm.residentialAccountId)
+                residences.extend(
+                    await self.hass.async_add_executor_job(account.get_residences)
```

Comment:
I know there are reasons why you should not create executor jobs in a for loop. Let me check this with someone else.

Reply:
Switching context to an executor job has quite a lot of overhead. So you would want to create a function where you do all the sync calls and then put it in the executor job

Reply:
makes sense, fixed

Reply:
Here
https://github.com/home-assistant/core/pull/104573/commits/de484465ca1614d335233bd54df04377d1af3b22